jax/core.py

```diff
@@ -811,14 +813,17 @@ def full_lower(val):
   else:
     return val
 
-def find_top_trace(xs) -> Trace:
+def find_top_trace(xs, axis_names=None) -> Trace:
+  top_main: Optional[MainTrace] = None
+  if axis_names:
```

Comment:
This feels a bit like the style of having two bits of functionality underneath one function, gated by configuration arguments, rather than just separating out two functions.

How about defining a new `bind` method on `AxisPrimitive`? That is, don't make any changes to `Primitive.bind` or to `core.find_top_trace`. Instead just have something like

```python
class AxisPrimitive(Primitive):
  def bind(self, *args, **params):
    top_trace = find_top_trace_with_axis_names(...)
    ...

def find_top_trace_with_axis_names(...):
  ...
```

We could even adapt the `bind` signature here to look like `def bind(self, *args, axis_name, **params):` or something, to be more explicit.

WDYT?

Reply:
(This also avoids the need for the `_dispatch_on_params` flag, as the type tag of `AxisPrimitive` has all the info we need.)

Reply:
I thought about it, but I disliked the code duplication. I am afraid that we'll end up changing those bits (e.g. because of your autodidax-inspired simplifications) and we'll just miss something important. Are you worried about the performance implications of this or is it something else?

Reply:
I'm not worried about performance; I just think it's easier to read if we don't pack functionality together like this. I don't think duplicating a few lines is a problem, and tests will easily catch any issues during a refactor.

Reply:
We can land this for now; no need to block. But I might revise it later!